pl_bolts/models/rl/ppo_model.py

```diff
@@ -0,0 +1,375 @@
+                # if trajectory ends abtruptly, boostrap value of next state
+                if (terminal or epoch_end) and not done:
+                    self.state = self.state.to(device=self.device)
+                    with torch.no_grad():
+                        _, _, value = self(self.state)
+                        last_value = value.item()
+                        steps_before_cutoff = self.episode_step
+                else:
+                    last_value = 0
+                    steps_before_cutoff = 0
+
+                # discounted cumulative reward
+                self.batch_qvals += self.discount_rewards(self.ep_rewards + [last_value], self.gamma)[:-1]
+                # advantage
+                self.batch_adv += self.calc_advantage(self.ep_rewards, self.ep_values, last_value)
+                # logs
+                self.epoch_rewards.append(sum(self.ep_rewards))
+                # reset params
+                self.ep_rewards = []
+                self.ep_values = []
+                self.episode_step = 0
+                self.state = torch.FloatTensor(self.env.reset())
+
+            if epoch_end:
+                train_data = zip(
+                    self.batch_states, self.batch_actions, self.batch_logp, self.batch_qvals, self.batch_adv
+                )
+
+                for state, action, logp_old, qval, adv in train_data:
+                    yield state, action, logp_old, qval, adv
+
+                self.batch_states.clear()
+                self.batch_actions.clear()
+                self.batch_adv.clear()
+                self.batch_logp.clear()
+                self.batch_qvals.clear()
+
+                # logging
+                self.avg_reward = sum(self.epoch_rewards) / self.steps_per_epoch
+
+                # if epoch ended abruptly, exlude last cut-short episode to prevent stats skewness
+                epoch_rewards = self.epoch_rewards
+                if not done:
+                    epoch_rewards = epoch_rewards[:-1]
+
+                total_epoch_reward = sum(epoch_rewards)
+                nb_episodes = len(epoch_rewards)
+
+                self.avg_ep_reward = total_epoch_reward / nb_episodes
+                self.avg_ep_len = (self.steps_per_epoch - steps_before_cutoff) / nb_episodes
+
+                self.epoch_rewards.clear()
+
+    def actor_loss(self, state, action, logp_old, adv) -> torch.Tensor:
+        pi, _ = self.actor(state)
+        logp = self.actor.get_log_prob(pi, action)
+        ratio = torch.exp(logp - logp_old)
+        clip_adv = torch.clamp(ratio, 1 - self.clip_ratio, 1 + self.clip_ratio) * adv
+        loss_actor = -(torch.min(ratio * adv, clip_adv)).mean()
+        return loss_actor
+
+    def critic_loss(self, state, qval) -> torch.Tensor:
+        value = self.critic(state)
+        loss_critic = (qval - value).pow(2).mean()
+        return loss_critic
+
+    def training_step(self, batch: Tuple[torch.Tensor, torch.Tensor], batch_idx, optimizer_idx):
```

Comment:
pls add test for training the model

Reply:
sure just added tests for training the model. although not sure how to add an assertion in the test function to test if training was complete

Reply:
@sid-sundrani I don't think we need to add any assertion here. I believe this is a test to make sure that there are no exceptions/errors raised during training.